src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp

```diff
@@ -130,24 +112,68 @@ void CompositionalMultiphaseBase::registerDataOnMesh( Group & meshBodies )
   DomainPartition const & domain = this->getGroupByPath< DomainPartition >( "/Problem/domain" );
   ConstitutiveManager const & cm = domain.getConstitutiveManager();
 
+  // 0. Find a "reference" fluid model name (at this point, models are already attached to subregions)
+  forMeshTargets( meshBodies, [&]( string const &,
+                                   MeshLevel & mesh,
+                                   arrayView1d< string const > const & regionNames )
+  {
+    mesh.getElemManager().forElementSubRegions( regionNames,
+                                                [&]( localIndex const,
+                                                     ElementSubRegionBase & subRegion )
```

Comment:
I'm not sure if it's useful for subregion loops in physics solvers, but in general when iterating over a set of objects from a given list, it might occasionally be useful to get an index back into the list, such as when there's other data associated with that list.

To reduce boilerplate, I've used the following tactic in a similar situation (it's on a [branch](https://github.com/GEOSX/GEOSX/blob/70df502864c39dc79888493a62be4ef233a32b5f/src/coreComponents/linearAlgebra/multiscale/MeshUtils.hpp#L182-L215) not yet merged): add a type trait (using our `IS_VALID_EXPRESSION` macro) to check whether user-provided function is callable with an index and object, or just an object, and call it accordingly. Not sure how it would affect compile times if added to such a ubiquitous class as `Group` though.

Reply:
If we want to reduce `boilerplate`, could it be possible to have something like
```c++
for( ElementSubRegionBase & subRegion: elemMgr.getElementSubRegions( regionNames ) ) {
   ...
}
```
or if we need an index
```c++
for( int i = 0; i<  elemMgr.numElementSubRegions( regionNames ); ++i ) {
   ElementSubRegionBase & subRegion = elemMgr.getElementSubRegion( i );
   ...
}
```
No (nested) `lambda`, no `traits`, no `callback`. Classical `for` loops.

Reply:
@klevzoff That is a neat idea!!

@TotoGaz In retrospect, there isn't much benefit to some of the `forBlahBlah` functions. Perhaps we can discuss the merit of them in a dev meeting.

Reply:
Actually, this _systematic_ `forXXX` pattern is very impactful since it hinders testability and makes abstraction much more complicated (if not impossible).
If you want I can prepare some presentation on what we can do so we can exchange on the subject when I come at LLNL.

Reply:
Ranges seems to be the most composable abstraction these days. With respect to the above two patterns, given a library like range-v3, it would be a matter of writing
```c++
elementManager.subregions() | for_each( [&]( ElementSubRegionBase & subRegion ) { ... } );
```
vs
```c++
elementManager.subregions() | enumerate | for_each( [&]( localIndex i, ElementSubRegionBase & subRegion ) { ... } );
```
(provided we implement the range returned by `elementManager.subregions()`, which shouldn't be too difficult)

Reply:
Well, as always it's not so simple. 🤷 

Obviously `ranges` look really convenient, and would at least end up to `chained actions` instead of `nested actions`.

My main concern would be: _"What will the effective type returned by `elementManager::subregions` be?"_
Because after that call, we can chain how we want or need to. Would it be a `lambda`, or even standard code...

The effective `range` types can become insane.
We would not like this returned type to reflect the internal implementation too much.
```c++
namespace rv = ranges::views;
auto r = rv::iota(0, 20)
       | rv::filter([](int i) { return i % 2 == 0; })
       | rv::transform([](int i){ return i * i; });
std::cout << abi::__cxa_demangle(typeid(r).name(), nullptr, nullptr, nullptr) << '\n';
```
prints `ranges::transform_view<ranges::filter_view<ranges::iota_view<int, int>, main::{lambda(int)#1}>, main::{lambda(int)#2}>`. What are we concretely going to return in our interfaces? If we want to support multiple implementations (or `mock`), we would need `ranges` that do not depend on the implementation.
I do not know if `ranges` are meant to be returned.

We could try to return the `any_view` range which does some type erasure. (It has a cost, but we'll never face the bottleneck for the kind of problems we are considering.) But, `any_view` is not in `c++20` and is not even proposed in `c++23` 😨 . So what are we going to decide? Are we going to commit to `ranges-v3` if we do not know if we'll be able to eventually replace it by the `std` implementations?

Maybe returning a good old `std::vector` of pointers is the best solution? 🤣 

(Plus I've read that compilation times of ranges are dumb. I've never tested in real projects anyway.)

Reply:
I have more thoughts on the subject, but I think the discussion has outgrown the PR comment :nerd_face: Too bad there is no magic button to create a new issue from a thread and automatically copy all the comments...

Back the the original comment, in short term I think we can consider removing the index parameter if there are no actual usages anywhere. Maybe in a follow-up PR though?

Reply:
We may want want to revive https://github.com/GEOSX/GEOSX/issues/1089, in which some elements (like `ranges`) were already mentioned?
Not exactly the same context, but I think it's the same idea.